crates/flight_datafusion/src/lib.rs

```diff
@@ -0,0 +1,330 @@
+#![allow(clippy::missing_errors_doc)]
+
+use arrow::array::RecordBatch;
+use async_stream::stream;
+use async_trait::async_trait;
+use flight_client::FlightClient;
+use futures::{Stream, StreamExt};
+use snafu::prelude::*;
+use sql_provider_datafusion::expr;
+use std::{any::Any, fmt, ops::Deref, pin::Pin, sync::Arc, task::Poll};
+use tokio::runtime::Handle;
+
+use arrow_flight::error::FlightError;
+use datafusion::{
+    arrow::datatypes::SchemaRef,
+    common::OwnedTableReference,
+    datasource::TableProvider,
+    error::{DataFusionError, Result as DataFusionResult},
+    execution::{context::SessionState, RecordBatchStream, TaskContext},
+    logical_expr::{Expr, TableProviderFilterPushDown, TableType},
+    physical_plan::{
+        project_schema, DisplayAs, DisplayFormatType, ExecutionPlan, SendableRecordBatchStream,
+    },
+    sql::TableReference,
+};
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Unable to generate SQL: {source}"))]
+    UnableToGenerateSQL { source: expr::Error },
+
+    #[snafu(display("Unable to query FlightSQL: {source}"))]
+    Flight { source: flight_client::Error },
+
+    #[snafu(display("Unable to query FlightSQL: {source}"))]
+    ArrowFlight { source: FlightError },
+
+    #[snafu(display("Unable to retrieve schema"))]
+    NoSchema,
+}
+
+type Result<T, E = Error> = std::result::Result<T, E>;
+
+pub struct FlightSQLTable {
+    client: Arc<FlightClient>,
+    schema: SchemaRef,
+    table_reference: OwnedTableReference,
+}
+
+impl FlightSQLTable {
+    pub fn new(
+        client: Arc<FlightClient>,
+        table_reference: impl Into<OwnedTableReference>,
+    ) -> Result<Self> {
+        let table_reference = table_reference.into();
+        let schema = Self::get_schema(Arc::clone(&client), &table_reference)?;
+        Ok(Self {
+            client,
+            schema,
+            table_reference,
+        })
+    }
+
+    #[allow(clippy::needless_pass_by_value)]
+    fn get_schema<'a>(
+        client: Arc<FlightClient>,
+        table_reference: impl Into<TableReference<'a>>,
+    ) -> Result<SchemaRef> {
+        let mut client = client.deref().clone();
+        tokio::task::block_in_place(move || {
+            Handle::current().block_on(async {
```

Comment:
We shouldn't do this - make `get_schema` and `new` async

Reply:
this means you will also need to make `get_table_provider` and `attach_mesh` async as well.